File: java/src/0001.java
import static org.assertj.core.api.Assertions.assertThat;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.HashMap;

class Solution {
    public int[] twoSum(int[] nums, int target) {
        var map = new HashMap<Integer, Integer>();
        for (int i = 0; i < nums.length; i++) {
            if (map.containsKey(target - nums[i])) {
                return new int[] { map.get(target - nums[i]), i };
            } else {
                map.put(nums[i], i);
            }
        }
        throw new IllegalArgumentException("Solution either does not exist or is not unique");
    }

    public static void main(String[] args) {
        var o = new Solution();

        assertThat(o.twoSum(new int[] { 2, 7, 11, 15 }, 9)).isEqualTo(new int[] { 0, 1 });
        assertThat(o.twoSum(new int[] { 3, 2, 4 }, 6)).isEqualTo(new int[] { 1, 2 });
        assertThat(o.twoSum(new int[] { 3, 3 }, 6)).isEqualTo(new int[] { 0, 1 });

        var start = Instant.now();
        for (var i = 0; i < 100000; i++) {
            o.twoSum(new int[] { 2, 7, 11, 15 }, 9);
            o.twoSum(new int[] { 3, 2, 4 }, 6);
            o.twoSum(new int[] { 3, 3 }, 6);
        }
        System.out.println(ChronoUnit.MICROS.between(start, Instant.now()));
    }
}

File: csharp/src/0001.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;

public class Solution {
    public int[] TwoSum(int[] nums, int target) {
        var m = new Dictionary<int, int>();
        for (var i = 0; i < nums.Length; i++) {
            if (m.ContainsKey(target - nums[i])) {
                return new[] { m[target - nums[i]], i };
            } else {
                m.Add(nums[i], i);
            }
        }
        throw new ArgumentException("Solution either does not exist or is not unique");
    }

    static void Main(string[] args) {
        var o = new Solution();

        Debug.Assert(o.TwoSum(new[] { 2, 7, 11, 15 }, 9).SequenceEqual(new[] { 0, 1 }));
        Debug.Assert(o.TwoSum(new[] { 3, 2, 4 }, 6).SequenceEqual(new[] { 1, 2 }));
        Debug.Assert(o.TwoSum(new[] { 3, 3 }, 6).SequenceEqual(new[] { 0, 1 }));

        var timer = new Stopwatch();
        timer.Start();
        for (var i = 0; i < 1000000; i++) {
            o.TwoSum(new[] { 2, 7, 11, 15 }, 9);
            o.TwoSum(new[] { 3, 2, 4 }, 6);
            o.TwoSum(new[] { 3, 3 }, 6);
        }
        timer.Stop();
        Console.WriteLine(timer.Elapsed.TotalMilliseconds);
    }
}
